File: pgpainless-core/src/main/java/org/pgpainless/key/generation/type/elgamal/ElGamal.java
// SPDX-FileCopyrightText: 2018 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation.type.elgamal;

import java.security.spec.AlgorithmParameterSpec;
import javax.annotation.Nonnull;

import org.bouncycastle.jce.spec.ElGamalParameterSpec;
import org.pgpainless.algorithm.PublicKeyAlgorithm;
import org.pgpainless.key.generation.type.KeyType;

/**
 * ElGamal encryption only key type.
 */
public final class ElGamal implements KeyType {

    private final ElGamalLength length;

    private ElGamal(@Nonnull ElGamalLength length) {
        this.length = length;
    }

    public static ElGamal withLength(ElGamalLength length) {
        return new ElGamal(length);
    }

    @Override
    public String getName() {
        return "ElGamal";
    }

    @Override
    public PublicKeyAlgorithm getAlgorithm() {
        return PublicKeyAlgorithm.ELGAMAL_ENCRYPT;
    }

    @Override
    public int getBitStrength() {
        return length.getLength();
    }

    @Override
    public AlgorithmParameterSpec getAlgorithmSpec() {
        return new ElGamalParameterSpec(length.getP(), length.getG());
    }

}

File: pgpainless-core/src/main/kotlin/org/pgpainless/key/generation/type/elgamal/ElGamal.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation.type.elgamal

import org.bouncycastle.jce.spec.ElGamalParameterSpec
import org.pgpainless.algorithm.PublicKeyAlgorithm
import org.pgpainless.key.generation.type.KeyType

/**
 * ElGamal encryption only key type.
 *
 * @deprecated the use of ElGamal is not recommended anymore.
 */
@Deprecated("The use of ElGamal is not recommended anymore.")
class ElGamal private constructor(length: ElGamalLength) : KeyType {

    override val name = "ElGamal"
    override val algorithm = PublicKeyAlgorithm.ELGAMAL_ENCRYPT
    override val bitStrength = length.length
    override val algorithmSpec = ElGamalParameterSpec(length.p, length.g)

    companion object {
        @JvmStatic
        fun withLength(length: ElGamalLength) = ElGamal(length)
    }
}